### PR TITLE
[SPARK-42908][PYTHON] Raise RuntimeError when SparkContext is required but not initialized

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Raise RuntimeError when SparkContext is required but not initialized.

### Why are the changes needed?
Error improvement.

### Does this PR introduce _any_ user-facing change?
Error type and message change. 

Raise a RuntimeError with a clear message (rather than an AssertionError) when SparkContext is required but not initialized yet.

### How was this patch tested?
Unit test.

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1211,7 +1211,10 @@ def _parse_datatype_string(s: str) -> DataType:
     from pyspark import SparkContext
 
     sc = SparkContext._active_spark_context
-    assert sc is not None
```
[the remark this answers is not in the document]
Comment (xinrong-meng) [reply python/pyspark/sql/types.py:1214]: Sounds good!

After the changes proposed in the PR, `python/pyspark/sql` has no such assertions.
```
 $ git grep 'assert sc is not None' -- python/pyspark/sql
 $ 
```

File: python/pyspark/sql/utils.py [hunk 2]
```diff
@@ -193,6 +193,15 @@ def wrapped(*args: Any, **kwargs: Any) -> Any:
     return cast(FuncT, wrapped)
 
 
+def require_spark_context_initialized() -> SparkContext:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/utils.py:196]: I would name it like: `get_jvm_spark_context`

**Review by HyukjinKwon — APPROVED** [review]
LGTM if tests pass

File: python/pyspark/sql/utils.py [hunk 3]
```diff
@@ -193,6 +193,15 @@ def wrapped(*args: Any, **kwargs: Any) -> Any:
...
+    """Raise RuntimeError if SparkContext is not initialized,
+    otherwise, returns the active SparkContext."""
+    sc = SparkContext._active_spark_context
+    if sc is None or sc._jvm is None:
+        raise RuntimeError("SparkContext must be initialized for JVM access.")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/utils.py:201]: I would just say something like: `SparkContext or SparkSession should be created first` instead

File: python/pyspark/sql/utils.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/sql/utils.py]: Good point, renamed.

File: python/pyspark/sql/utils.py (hunk 3 above)
Comment (xinrong-meng) [reply python/pyspark/sql/utils.py]: Updated.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Pull request closed without merging**